extensions/openstack/openstack-impl/src/main/java/org/mqnaas/extensions/openstack/capabilities/impl/OpenstackHostAdministration.java

```diff
@@ -0,0 +1,182 @@
+
+		try {
+			novaClient = clientProviderFactory.getClientProvider(IJCloudsNovaClientProvider.class).getClient(resource);
+		} catch (EndpointNotFoundException e) {
+			throw new ApplicationActivationException("Could not instantiate JClouds client.", e);
+		} catch (ProviderNotFoundException e) {
+			throw new ApplicationActivationException("Could not instantiate JClouds client.", e);
+		}
+
+		log.info("Initialized OpenstackHostAdministration capability for resource " + resource.getId());
+
+	}
+
+	@Override
+	public void deactivate() {
+		log.info("Removing OpenstackHostAdministration capability from resource " + resource.getId());
+
+		try {
+			Closeables.close(novaClient, true);
+		} catch (IOException e) {
+			log.warn("Could not close jclouds nova client.", e);
+		}
+
+		log.info("Removed OpenstackHostAdministration capability from resource " + resource.getId());
+
+	}
+
+	@Override
+	public int getNumberOfCpus() {
+		log.debug("Getting number of cpus of host [id=" + resource.getId() + "]");
+
+		Server server = getServer();
+		Flavor flavor = (Flavor) server.getFlavor();
+
+		return flavor.getVcpus();
+	}
+
+	@Override
+	public int getMemorySize() {
+		log.debug("Getting memory size of host [id=" + resource.getId() + "]");
+
+		Server server = getServer();
+		Flavor flavor = (Flavor) server.getFlavor();
+
+		return flavor.getRam();
+	}
+
+	@Override
+	public int getDiskSize() {
+		log.debug("Getting disk size of host [id=" + resource.getId() + "]");
+
+		Server server = getServer();
+		Flavor flavor = (Flavor) server.getFlavor();
+
+		return flavor.getDisk();
+	}
+
+	@Override
+	public String getSwapSize() {
+
+		Server server = getServer();
+		Flavor flavor = (Flavor) server.getFlavor();
+
+		return flavor.getSwap() != null ? flavor.getSwap().get() : null;
+	}
+
+	/**
+	 * Retrieves the {@link Server} instance represented by the injected <code>resource</code> by using the jclouds client.
+	 * 
+	 * @throws IllegalStateException
+	 *             <ul>
+	 *             <li>If {@link IAttributeStore} capability of the injected resource does not contain information about the {@link Server} id and the
+	 *             zone it belongs to.</li>
+	 *             <li>If there's no zone identified by the zoneId stored in IAttributeStore capability.</li>
+	 *             <li>If there's no server identified by the id stored in IAttributeStore capability.</li>
+	 *             </ul>
+	 */
+	private Server getServer() {
+
+		String zone = attributeStore.getAttribute(OpenstackRootResourceProvider.ZONE_ATTRIBUTE);
+		String vmId = attributeStore.getAttribute(IAttributeStore.RESOURCE_EXTERNAL_ID);
+
+		if (StringUtils.isEmpty(zone) || StringUtils.isEmpty(vmId))
+			throw new IllegalStateException("Can't read VM cpus if AttributeStore does not contain its external id and the zone it belongs to.");
+
+		ServerApi serverClient = novaClient.getServerApiForZone(zone);
```

Comment:
Reusing a client may not have desired behaviour. Probably credentials will not be set.
Is this on purpose?